log/sync.go

```diff
@@ -0,0 +1,233 @@
+}
+
+// NewSyncWriter returns a new SyncWriter. The returned writer is safe for
+// concurrent use by multiple goroutines.
+func NewSyncWriter(w io.Writer) *SyncWriter {
+	return &SyncWriter{w: w}
+}
+
+// Write writes p to the underlying io.Writer. If another write is already in
+// progress, the calling goroutine blocks until the SyncWriter is available.
+func (w *SyncWriter) Write(p []byte) (n int, err error) {
+	w.mu.Lock()
+	n, err = w.w.Write(p)
+	w.mu.Unlock()
+	return n, err
+}
+
+// syncLogger provides concurrent safe logging for another Logger.
+type syncLogger struct {
+	mu     sync.Mutex
+	logger Logger
+}
+
+// NewSyncLogger returns a logger that synchronizes concurrent use of the
+// wrapped logger. When multiple goroutines use the SyncLogger concurrently
+// only one goroutine will be allowed to log to the wrapped logger at a time.
+// The other goroutines will block until the logger is available.
+func NewSyncLogger(logger Logger) Logger {
+	return &syncLogger{logger: logger}
+}
+
+// Log logs keyvals to the underlying Logger. If another log is already in
+// progress, the calling goroutine blocks until the syncLogger is available.
+func (l *syncLogger) Log(keyvals ...interface{}) error {
+	l.mu.Lock()
+	err := l.logger.Log(keyvals...)
+	l.mu.Unlock()
+	return err
+}
+
+// AsyncLogger provides buffered asynchronous and concurrent safe logging for
+// another logger.
+//
+// Errors returned by the wrapped logger are ignored, therefore the wrapped
+// logger must handle all errors appropriately.
+type AsyncLogger struct {
+	logger   Logger
+	keyvalsC chan []interface{}
+}
+
+// NewAsyncLogger returns a new AsyncLogger that logs to logger and can buffer
+// up to size log events before its Log method blocks.
+func NewAsyncLogger(logger Logger, size int) *AsyncLogger {
+	l := &AsyncLogger{
+		logger:   logger,
+		keyvalsC: make(chan []interface{}, size),
+	}
+	go l.run()
+	return l
+}
+
+// run forwards log events from l.keyvalsC to l.logger.
+func (l *AsyncLogger) run() {
+	for keyvals := range l.keyvalsC {
+		l.logger.Log(keyvals...)
+	}
+}
+
+// Log queues keyvals for logging by the wrapped Logger. Log may be called
+// concurrently by multiple goroutines. If the the buffer is full, Log will
+// block until space is available. Log always returns a nil error.
+func (l *AsyncLogger) Log(keyvals ...interface{}) error {
+	l.keyvalsC <- keyvals
+	return nil
+}
+
+// Len returns a snapshot of the number of buffered log events. The returned
+// count should only be used for monitoring purposes as it becomes stale
+// quickly.
+func (l *AsyncLogger) Len() int {
+	return len(l.keyvalsC)
+}
+
+// Cap returns the maximum capacity of the buffer.
+func (l *AsyncLogger) Cap() int {
+	return cap(l.keyvalsC)
+}
+
+// NonblockingLogger provides buffered asynchronous and concurrent safe
+// logging for another logger.
+//
+// Errors returned by the wrapped logger are ignored, therefore the wrapped
+// logger must handle all errors appropriately.
```

Comment:
Your first two paragraphs are spot on.

What value does having a `StopOnError` option provide? In what scenario would anyone enable that option and not want to also handle the error?